### ISSUE TITLE
Separate Kibana instance won’t be able to connect to Elasticsearch on localhost

### ISSUE THREAD

**Issue by exalted** [spujadas/elk-docker#199]
Running Docker CE Version 17.12.0-ce-mac55 (23011) for Mac and set Memory to 4.0 GB.

When using this image with the following `docker-compose.yml`:

```yaml
version: '2'
services:
  logs:
    image: sebp/elk:622

    environment:
      - MAX_MAP_COUNT=262144

    ports:
      - 5601:5601 # Kibana web interface
      - 9200:9200 # Elasticsearch JSON interface
      - 5044:5044 # Logstash Beats interface

    volumes:
      - /var/lib/elasticsearch
```

Everything seem to work just fine. However, if I had to create separate `docker-compose.yml` files (one for Elasticsearch & Logstash and one for Kibana only):

```yaml
version: '2'
services:
  # back end (Elasticsearch & Logstash)
  logs:
    image: sebp/elk:622

    environment:
      - MAX_MAP_COUNT=262144
      - KIBANA_START=0

    ports:
      - 9200:9200 # Elasticsearch JSON interface
      - 5044:5044 # Logstash Beats interface

    volumes:
      - /var/lib/elasticsearch
```

```yaml
version: '2'
services:
  # front end (Kibana)
  logs:
    image: sebp/elk:622

    environment:
      - ELASTICSEARCH_START=0
      - LOGSTASH_START=0

    ports:
      - 5601:5601 # Kibana web interface
```

Kibana will complain like so:

```
logs   │ {"type":"log","@timestamp":"2018-03-19T16:31:55Z","tags":["warning","elasticsearch","admin"],"pid":28,"message":"No living connections"}
logs   │ {“type":"log","@timestamp":"2018-03-19T16:31:58Z","tags":["warning","elasticsearch","admin"],"pid":28,"message":"Unable to revive connection: http://localhost:9200/"}
```

However `curl -XGET ‘http://localhost:9200'` will respond with:

```json
{
  "name" : "s0WtIrY",
  "cluster_name" : "elasticsearch",
  "cluster_uuid" : "Vmbec-rISyWmjhPJVsYnyQ",
  "version" : {
    "number" : "6.2.2",
    "build_hash" : "10b1edd",
    "build_date" : "2018-02-16T19:01:30.685723Z",
    "build_snapshot" : false,
    "lucene_version" : "7.2.1",
    "minimum_wire_compatibility_version" : "5.6.0",
    "minimum_index_compatibility_version" : "5.0.0"
  },
  "tagline" : "You Know, for Search"
}
```

Also, I can create a dummy log entry like so: `curl -XPUT ‘http://localhost:9200/dummyindex-today/dummytype/foo' -H ‘Content-Type: application/json’ -d ‘{ “name” : “Dilbert Brown”, “user”: "dilbert", "postDate": "2011-12-15" }'`, so that curl -XGET 'http://localhost:9200/dummyindex-today' will reply:

```json
{
  "aliases": {},
  "mappings": {
    "dummytype": {
      "properties": {
        "name": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        },
        "postDate": {
          "type": "date"
        },
        "user": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        }
      }
    }
  },
  "settings": {
    "index": {
      "creation_date": "1521474283229",
      "number_of_shards": "5",
      "number_of_replicas": "1",
      "uuid": "tqzVdLUcRwq7Se_676yOzg",
      "version": {
        "created": "6020299"
      },
      "provided_name": "dummyindex-today"
    }
  }
}
```

What am I missing here? Why do you think although Elasticsearch seems to be working just fine on http://localhost:9200, for some reason Kibana is not able to “connect” to it?

Thanks.

**Comment by spujadas** [comment]
_tl;dr – Look into Docker networking for information on making containers communicate with each other._

Two things here:

- The two containers publish their ports to the host but are actually "isolated" from each other, i.e. you can access both of them from the host but (by default, especially – I think – when running the containers through two separate instances of `docker-compose`) they don't know about each other.

- The default configuration files in the image assume that all the services are running locally within the container, i.e. that they're all running on _localhost_ (as seen from the container, i.e. the container itself, not the host – hope that makes sense). In other words, Kibana attempts to connect to Elasticsearch on _localhost_ within the same container… but fails as Elasticsearch is actually not running there.

An easy (but dirty) way (that won't work in all instances) to make this work is to update Kibana's configuration file (`kibana.yml`) to make `elasticsearch.url` point to an IP address of your host that is routable from within the Kibana-running container.
The proper way to manage this is to set up networking for Compose properly. Start your journey here: https://docs.docker.com/compose/networking/

**Comment by exalted** [comment]
> The default configuration files in the image assume that all the services are running locally within the container, i.e. that they're all running on localhost (as seen from the container, i.e. the container itself, not the host – hope that makes sense). In other words, Kibana attempts to connect to Elasticsearch on localhost within the same container… but fails as Elasticsearch is actually not running there.

D-Oh! 🤦‍♂️ I must have gone out of my mind asking this. Thanks for the _wake up!_ call…

> An easy (but dirty) way (that won’t work in all instances) to make this work is to update Kibana's configuration file (kibana.yml) to make elasticsearch.url point to an IP address of your host that is routable from within the Kibana-running container.

Side note: on Docker for Mac there’s also `docker.for.mac.internal` (previously `docker.for.mac.localhost`). 

Thanks @spujadas ! 🙇

**Comment by spujadas** [comment]
No problem, and thanks for the side note! :+1:
